tensorboard/plugins/mesh/tf_mesh_dashboard/mesh-loader.js

```diff
@@ -167,11 +162,34 @@ Polymer({
    * @private
    */
   _updateScene: function(currentStep) {
-    this._meshViewer.updateScene(currentStep, this);
-    if (!this._cameraPositionInitialized) {
-      this._meshViewer.resetView();
-      this._cameraPositionInitialized = true;
+    let dataPromise = Promise.resolve();
+    if (!currentStep.mesh) {  // Lazy-load actual mesh data.
+      dataPromise = this._dataProvider.fetchData(
+        currentStep, this.run, this.tag, this.sample, this._stepIndex)
+      .then(function (meshData) {
+        currentStep.mesh = meshData[0];
+      })
+      .catch((error) => {
+        if (!error || !error.code || error.code != vz_mesh.ErrorCodes.CANCELLED) {
+          error = error || 'Response processing failed.';
+          throw new Error(error);
+        }
+      });;
     }
+
+    return dataPromise.then(function () {
+      this._meshViewer.updateScene(currentStep, this);
+      if (!this._cameraPositionInitialized) {
+        this._meshViewer.resetView();
+        this._cameraPositionInitialized = true;
+      }
+      this.set('_isMeshLoading', false);
```

Comment:
IMO, this is quite confusing. So AFAICT, this change first fetches the `steps` and lazy loads the mesh of a `step` depending on what is currently active. There are two separate request/loading involved and `_updateScene` can get triggered multiple times when user uses the step slider to change the `currentStep`. In my opinion, there should be a two separate loading state potentially with a helpful text for users. Two loading states especially makes sense since user will move the slider and sees no UI change until the mesh is fetched. 

Separate refactor idea: I think `updateScene` should not have any fetching logic but simply update meshViewer when `currentStep.mesh` is present. Another observer can listen to the `currentStep` that fetches the mesh if it does not exist.

Reply:
How can I guarantee order in  each observers are called and that data is fetched before `updateScene`?

Reply:
Also, loading indicator will be displayed when mesh is loading (as well as metadata), so I don't see why should we have two separate states. Especially considering that they will never overlap.